File: backend/app/models/enums.py
"""Enumerations metier.

Elles sont mappees en VARCHAR + CHECK (`native_enum=False`) et non en type ENUM
PostgreSQL natif, pour deux raisons :

1. ajouter une valeur a un ENUM natif est une migration bloquante penible ;
2. SQLite -- donc Drift, cote tablette -- ne connait pas les ENUM. Un VARCHAR
   contraint donne exactement le meme schema logique des deux cotes, ce qui est
   la condition pour que la synchronisation reste un simple mapping 1:1.
"""

from __future__ import annotations

import enum


class StrEnum(str, enum.Enum):
    def __str__(self) -> str:
        return self.value


# --- Chambres ---------------------------------------------------------------


class OccupancyStatus(StrEnum):
    VACANT = "VACANT"
    RESERVED = "RESERVED"
    OCCUPIED = "OCCUPIED"


class HousekeepingStatus(StrEnum):
    CLEAN = "CLEAN"
    DIRTY = "DIRTY"
    IN_PROGRESS = "IN_PROGRESS"
    INSPECTED = "INSPECTED"


class TaxMode(StrEnum):
    PERCENT = "PERCENT"
    PER_NIGHT = "PER_NIGHT"
    PER_PERSON_NIGHT = "PER_PERSON_NIGHT"
    FIXED = "FIXED"


# --- Reservations -----------------------------------------------------------


class ReservationStatus(StrEnum):
    PENDING = "PENDING"
    CONFIRMED = "CONFIRMED"
    CHECKED_IN = "CHECKED_IN"
    CHECKED_OUT = "CHECKED_OUT"
    CANCELLED = "CANCELLED"
    NO_SHOW = "NO_SHOW"


class ReservationSource(StrEnum):
    DIRECT = "DIRECT"
    PHONE = "PHONE"
    WALK_IN = "WALK_IN"
    OTA = "OTA"
    CORPORATE = "CORPORATE"
    EMAIL = "EMAIL"


class SignatureKind(StrEnum):
    CHECK_IN = "CHECK_IN"
    CHECK_OUT = "CHECK_OUT"
    INVOICE = "INVOICE"
    OTHER = "OTHER"


# --- Facturation ------------------------------------------------------------


class FolioType(StrEnum):
    GUEST = "GUEST"
    MASTER = "MASTER"
    WALK_IN = "WALK_IN"
    TABLE = "TABLE"


class FolioStatus(StrEnum):
    OPEN = "OPEN"
    CLOSED = "CLOSED"
    SETTLED = "SETTLED"


class ChargeCategory(StrEnum):
    ROOM = "ROOM"
    FNB = "FNB"
    MINIBAR = "MINIBAR"
    SPA = "SPA"
    LAUNDRY = "LAUNDRY"
    TELEPHONE = "TELEPHONE"
    TAX = "TAX"
    DISCOUNT = "DISCOUNT"
    DEPOSIT = "DEPOSIT"
    MISC = "MISC"


class InvoiceStatus(StrEnum):
    DRAFT = "DRAFT"
    ISSUED = "ISSUED"
    PAID = "PAID"
    CANCELLED = "CANCELLED"


class PaymentMethod(StrEnum):
    CASH = "CASH"
    CARD = "CARD"
    TRANSFER = "TRANSFER"
    MOBILE_MONEY = "MOBILE_MONEY"
    CITY_LEDGER = "CITY_LEDGER"
    VOUCHER = "VOUCHER"


class CashSessionStatus(StrEnum):
    OPEN = "OPEN"
    CLOSED = "CLOSED"


# --- Restauration -----------------------------------------------------------


class OrderType(StrEnum):
    """Mode de service, et lui seul.

    Le *lieu* est porte par `orders.outlet_id` (restaurant, bar, piscine,
    boite de nuit). Melanger les deux -- l'ancienne valeur `POOL` etait un
    lieu -- rendait la piscine representable de deux facons, ce qui garantit
    de retrouver les deux dans les donnees.

    Une consommation au bord du bassin est donc ON_SITE / outlet POOL, et une
    commande de la boite de nuit montee en chambre devient ROOM_SERVICE /
    outlet NIGHTCLUB -- combinaison inexprimable auparavant.
    """

    ON_SITE = "ON_SITE"
    ROOM_SERVICE = "ROOM_SERVICE"
    TAKEAWAY = "TAKEAWAY"
    DELIVERY = "DELIVERY"


class OrderStatus(StrEnum):
    DRAFT = "DRAFT"
    SENT = "SENT"
    IN_PREP = "IN_PREP"
    READY = "READY"
    SERVED = "SERVED"
    CANCELLED = "CANCELLED"


class TableStatus(StrEnum):
    FREE = "FREE"
    OCCUPIED = "OCCUPIED"
    RESERVED = "RESERVED"


# --- Housekeeping / maintenance ---------------------------------------------


class HousekeepingTaskType(StrEnum):
    DEPARTURE = "DEPARTURE"
    STAYOVER = "STAYOVER"
    REFRESH = "REFRESH"
    DEEP_CLEAN = "DEEP_CLEAN"
    INSPECTION = "INSPECTION"


class TaskStatus(StrEnum):
    PENDING = "PENDING"
    ASSIGNED = "ASSIGNED"
    IN_PROGRESS = "IN_PROGRESS"
    DONE = "DONE"
    INSPECTED = "INSPECTED"
    CANCELLED = "CANCELLED"


class Priority(StrEnum):
    LOW = "LOW"
    NORMAL = "NORMAL"
    HIGH = "HIGH"
    URGENT = "URGENT"


class TicketStatus(StrEnum):
    OPEN = "OPEN"
    ASSIGNED = "ASSIGNED"
    IN_PROGRESS = "IN_PROGRESS"
    RESOLVED = "RESOLVED"
    CLOSED = "CLOSED"
    CANCELLED = "CANCELLED"


# --- Stocks -----------------------------------------------------------------


class StockMovementType(StrEnum):
    IN = "IN"
    OUT = "OUT"
    TRANSFER = "TRANSFER"
    ADJUSTMENT = "ADJUSTMENT"
    LOSS = "LOSS"
    RETURN = "RETURN"


class InventoryStatus(StrEnum):
    DRAFT = "DRAFT"
    COUNTING = "COUNTING"
    CLOSED = "CLOSED"


# --- Impression -------------------------------------------------------------


class PrinterKind(StrEnum):
    LASER = "LASER"
    THERMAL = "THERMAL"


class PrinterProtocol(StrEnum):
    IPP = "IPP"
    LPD = "LPD"
    ESCPOS_NET = "ESCPOS_NET"
    ESCPOS_USB = "ESCPOS_USB"
    RAW9100 = "RAW9100"


class PrintJobStatus(StrEnum):
    QUEUED = "QUEUED"
    SENT = "SENT"
    PRINTED = "PRINTED"
    FAILED = "FAILED"
    CANCELLED = "CANCELLED"


class TemplateFormat(StrEnum):
    ESCPOS = "ESCPOS"
    HTML = "HTML"
    PDF = "PDF"


# --- Synchronisation --------------------------------------------------------


class SyncOp(StrEnum):
    INSERT = "INSERT"
    UPDATE = "UPDATE"
    DELETE = "DELETE"


class ConflictResolution(StrEnum):
    SERVER_WINS = "SERVER_WINS"
    CLIENT_WINS = "CLIENT_WINS"
    MERGED = "MERGED"
    MANUAL = "MANUAL"
    PENDING = "PENDING"


class UploadState(StrEnum):
    PENDING = "PENDING"
    UPLOADING = "UPLOADING"
    UPLOADED = "UPLOADED"
    FAILED = "FAILED"


# --- Divers -----------------------------------------------------------------


class SettingScope(StrEnum):
    GLOBAL = "GLOBAL"
    DEVICE = "DEVICE"
    USER = "USER"


class BusinessDayStatus(StrEnum):
    OPEN = "OPEN"
    CLOSED = "CLOSED"


class IdDocumentType(StrEnum):
    ID_CARD = "ID_CARD"
    PASSPORT = "PASSPORT"
    DRIVING_LICENSE = "DRIVING_LICENSE"
    RESIDENCE_PERMIT = "RESIDENCE_PERMIT"
    OTHER = "OTHER"

File: frontend/lib/data/local/enums.dart
/// Enumerations metier de la base locale.
///
/// Elles sont le miroir exact de `backend/app/models/enums.py`. Chaque valeur
/// est stockee sous forme de texte (`textEnum`), avec la meme chaine des deux
/// cotes : c'est ce qui permet a la synchronisation de recopier la valeur telle
/// quelle, sans table de correspondance ni conversion.
///
/// Consequence importante : ne jamais renommer une valeur existante, et ne
/// jamais se fier a l'ordre de declaration. Ajouter une valeur est sans risque,
/// en renommer une casse toutes les lignes deja enregistrees.
library;

// --- Chambres ---------------------------------------------------------------

enum OccupancyStatus { VACANT, RESERVED, OCCUPIED }

enum HousekeepingStatus { CLEAN, DIRTY, IN_PROGRESS, INSPECTED }

enum TaxMode { PERCENT, PER_NIGHT, PER_PERSON_NIGHT, FIXED }

/// Pastille affichee sur le plan de l'hotel (ecran 5.2 du cahier des charges).
///
/// Cette valeur n'est pas stockee : elle se calcule a partir des trois axes
/// independants d'une chambre. Voir `Room.displayStatus`.
enum RoomDisplayStatus { AVAILABLE, OCCUPIED, RESERVED, CLEANING, MAINTENANCE }

// --- Reservations -----------------------------------------------------------

enum ReservationStatus {
  PENDING,
  CONFIRMED,
  CHECKED_IN,
  CHECKED_OUT,
  CANCELLED,
  NO_SHOW,
}

enum ReservationSource { DIRECT, PHONE, WALK_IN, OTA, CORPORATE, EMAIL }

enum SignatureKind { CHECK_IN, CHECK_OUT, INVOICE, OTHER }

// --- Facturation ------------------------------------------------------------

enum FolioType { GUEST, MASTER, WALK_IN, TABLE }

enum FolioStatus { OPEN, CLOSED, SETTLED }

enum ChargeCategory {
  ROOM,
  FNB,
  MINIBAR,
  SPA,
  LAUNDRY,
  TELEPHONE,
  TAX,
  DISCOUNT,
  DEPOSIT,
  MISC,
}

enum InvoiceStatus { DRAFT, ISSUED, PAID, CANCELLED }

enum PaymentMethod { CASH, CARD, TRANSFER, MOBILE_MONEY, CITY_LEDGER, VOUCHER }

enum CashSessionStatus { OPEN, CLOSED }

// --- Restauration -----------------------------------------------------------

/// Mode de service, et lui seul : le *lieu* est porte par `orders.outletId`.
///
/// L'ancienne valeur `POOL` etait un lieu, pas un mode -- la piscine etait
/// donc representable de deux facons, ce qui garantit de retrouver les deux
/// dans les donnees. Une consommation au bord du bassin est desormais
/// ON_SITE / outlet POOL.
enum OrderType { ON_SITE, ROOM_SERVICE, TAKEAWAY, DELIVERY }

enum OrderStatus { DRAFT, SENT, IN_PREP, READY, SERVED, CANCELLED }

enum TableStatus { FREE, OCCUPIED, RESERVED }

// --- Housekeeping et maintenance --------------------------------------------

enum HousekeepingTaskType {
  DEPARTURE,
  STAYOVER,
  REFRESH,
  DEEP_CLEAN,
  INSPECTION,
}

enum TaskStatus { PENDING, ASSIGNED, IN_PROGRESS, DONE, INSPECTED, CANCELLED }

enum Priority { LOW, NORMAL, HIGH, URGENT }

enum TicketStatus { OPEN, ASSIGNED, IN_PROGRESS, RESOLVED, CLOSED, CANCELLED }

// --- Stocks -----------------------------------------------------------------

enum StockMovementType { IN, OUT, TRANSFER, ADJUSTMENT, LOSS, RETURN }

enum InventoryStatus { DRAFT, COUNTING, CLOSED }

// --- Impression -------------------------------------------------------------

enum PrinterKind { LASER, THERMAL }

enum PrinterProtocol { IPP, LPD, ESCPOS_NET, ESCPOS_USB, RAW9100 }

enum PrintJobStatus { QUEUED, SENT, PRINTED, FAILED, CANCELLED }

enum TemplateFormat { ESCPOS, HTML, PDF }

// --- Synchronisation --------------------------------------------------------

enum SyncOp { INSERT, UPDATE, DELETE }

enum SyncState { synced, pending, conflict }

enum OutboxStatus { PENDING, SENDING, ACKED, FAILED }

enum UploadState { PENDING, UPLOADING, UPLOADED, FAILED }

// --- Divers -----------------------------------------------------------------

enum SettingScope { GLOBAL, DEVICE, USER }

enum BusinessDayStatus { OPEN, CLOSED }

enum IdDocumentType {
  ID_CARD,
  PASSPORT,
  DRIVING_LICENSE,
  RESIDENCE_PERMIT,
  OTHER,
}
